llvm/include/llvm/Analysis/ScalarEvolution.h

```diff
@@ -898,6 +901,13 @@ class ScalarEvolution {
     return getBackedgeTakenCount(L, ConstantMaximum);
   }
 
+  /// Similar to getConstantMaxBackedgeTakenCount, except it will add a set of
+  /// SCEV predicates to Predicates that are required to be true in order for
+  /// the answer to be correct. Predicates can be checked with run-time
+  /// checks and can be used to perform loop versioning.
+  const SCEV *getPredicatedConstantMaxBackedgeTakenCount(
```

Comment:
Should those also be included in SCEV printing? Maybe it already is, and maybe just a SCEV test is missing?

Reply:
Good point and thanks for reminding me!

Reply:
With the printing change, I suppose adding the new SCEV apis would become independent of updating  isDereferenceableAndAlignedInLoop

Reply:
True, but it feels a bit unnecessary/artificial to split the patch up given the main motivation for this work is precisely to improve code quality from the vectoriser - see the tests in Transforms/LoopVectorize/load-deref-pred-align.ll. It will obviously benefit PR #88385 too, but as a standalone patch it makes some existing vectorised loops better.

Reply:
Having said that, if you really prefer me to split the patch up because you anticipate potential fall-out from the change in vectoriser behaviour I can do that. I guess I thought the change seemed unlikely to cause problems.